llpc/context/llpcCompiler.cpp

```diff
@@ -602,66 +849,9 @@ Result Compiler::buildPipelineWithRelocatableElf(Context *context, ArrayRef<cons
-    const auto shaderStages = maskToShaderStages(shaderStageMask);
-    assert(all_of(shaderStages, isNativeStage) && "Unexpected stage kind");
-
-    // Check the cache for the relocatable shader for this stage .
-    MetroHash::Hash cacheHash = {};
-    if (context->isGraphics()) {
-      auto pipelineInfo = reinterpret_cast<const GraphicsPipelineBuildInfo *>(context->getPipelineBuildInfo());
-      cacheHash = PipelineDumper::generateHashForGraphicsPipeline(pipelineInfo, true, true, stage);
-    } else {
-      auto pipelineInfo = reinterpret_cast<const ComputePipelineBuildInfo *>(context->getPipelineBuildInfo());
-      cacheHash = PipelineDumper::generateHashForComputePipeline(pipelineInfo, true, true);
-    }
-    // Note that this code updates m_pipelineHash of the pipeline context. It
-    // must be restored before we link the pipeline ELF at the end of this for-loop.
-    context->getPipelineContext()->setHashForCacheLookUp(cacheHash);
-    LLPC_OUTS("Finalized hash for " << getUnlinkedShaderStageName(stage) << " stage cache lookup: "
-                                    << format_hex(context->getPipelineContext()->get128BitCacheHashCode()[0], 18) << ' '
-                                    << format_hex(context->getPipelineContext()->get128BitCacheHashCode()[1], 18)
-                                    << '\n');
-
-    CacheAccessor cacheAccessor(context, cacheHash, getInternalCaches());
-    if (cacheAccessor.isInCache()) {
-      BinaryData elfBin = cacheAccessor.getElfFromCache();
-      auto data = reinterpret_cast<const char *>(elfBin.pCode);
-      elf[stage].assign(data, data + elfBin.codeSize);
-      LLPC_OUTS("Cache hit for shader stage " << getUnlinkedShaderStageName(stage) << "\n");
-      for (ShaderStage stage : shaderStages)
-        stageCacheAccesses[stage] =
-            cacheAccessor.hitInternalCache() ? CacheAccessInfo::InternalCacheHit : CacheAccessInfo::CacheHit;
-
-      continue;
-    }
-
-    LLPC_OUTS("Cache miss for shader stage " << getUnlinkedShaderStageName(stage) << "\n");
-    for (ShaderStage stage : shaderStages)
-      stageCacheAccesses[stage] = CacheAccessInfo::CacheMiss;
-
-    // There was a cache miss, so we need to build the relocatable shader for
-    // this stage.
-    const PipelineShaderInfo *singleStageShaderInfo[ShaderStageNativeStageCount] = {nullptr, nullptr, nullptr,
-                                                                                    nullptr, nullptr, nullptr};
-    for (ShaderStage stage : shaderStages)
-      singleStageShaderInfo[stage] = shaderInfo[stage];
-
-    Vkgc::ElfPackage &stageElf = elf[stage];
-    result = buildPipelineInternal(context, singleStageShaderInfo, PipelineLink::Unlinked, nullptr, &stageElf,
-                                   stageCacheAccesses);
-    if (result != Result::Success)
+    if (buildUnlinkedShaderInternal(context, shaderInfo, stage, elf[stage], stageCacheAccesses) != Result::Success)
```

Comment:
I like this.  This function was too long.  It helps readability as well as code reuse.